### ISSUE TITLE
Include value binding features to DriverInterface

### ISSUE THREAD

**Issue by dmill-bz** [spider/spider#23]
Binding values to queries in order to prevent injections is going to be a common feature of all databases and driver clients. We should add functionality around these in `DriverInterface`

Something along the lines of:

``` php
/**
 * @return void
 */
public function bindValue(string $name, $value);

/**
 * @return void
 */
public function bindValues(array $binds);

/**
 * @return array (key/value)
 */
public function getBinds();
```

The discussion around binds can be much larger. We will need to settle on a consistent bindpattern (`{var}`, `B_VAR`, etc.) for `Command` and then it would be the `CommandProcessor`'s job to correctly format the binds to whichever language and pass them to the driver.

**Comment by electricjones** [comment]
It should be possible to use the diver without the CommandProcessor if you know what you are doing.

To that end, what if the `Command` holds the script, the script language, and the values to bind? The driver then, when `executeXCommand()` is fired, binds the values to the statement before sending it along.

``` php
$driver = new OrientDriver($credentials)

$values = [
    'name' => 'Malcom Reynolds',
    'ship' => 'Serenity'
];

$command = new Command("INSERT INTO Profile SET {{values}}", $values);
// or $command->setValues($values);
```

The `Driver` then receives the `Command` and performs the correct cleanup and regex replacements.

``` php
public function executeWriteCommand($command) {
    $cleanCommand = // get the values from the Command, sanitize them, and regex replace
    // Becomes INSERT INTO Profile SET name = 'Malcom Reynolds', ship= 'Serenity'
    $response = // send it all to the database
}
```

I don't know if I like this, though, because it gives the driver responsibility in writing query statements. 

It may be an _additional_ option to binding in the `Driver` itself.

**Comment by dmill-bz** [comment]
I've got a counter-proposal that is almost the same thing as what you suggested.
It works on the basis that:
- Every `Processor` should have functionality around cleaning and regex replacements of binds. Or at least extend a class that does.
- Knowledge of the format of binds for it's language. `{{value}}` | `:value` | etc. (we will have to settle on something if there is no specific format for a language)
- Return a `new Command('query string', 'language', ['bindKey' => 'bindValue'])`

With this there are two scenarios:
- The `Driver` has a specific tunnel for binds (ie: the server handles the bindings and ensures correct escaping/injection-protection). In which case the `Driver` will just add the binds separately to the request:
  
  ``` php
  $driver = new GremlinDriver($credentials)
  
  $values = [
  'B_NAME' => 'Malcom Reynolds',
  'B_SHIP' => 'Serenity'
  ];
  
  $command = new Command("g.V().has('name', B_NAME).has('ship', B_SHIP)", 'gremlin-groovy', $values);
  ```
  
  Works well. we just need to make sure that the Processors force the expected bind format `{{values}}`, `B_VALUES`, `:values` etc..
- The `Driver` does not pass the binds to the server. It should know which languages it can accept and should be capable of loading the appropriate `Processor` for the language provided in the `Command` (we already have `makeProcess()`). It then gives the `Command` to the `Processor` for bind cleaning/replacement and obtains a new bindless `Command`

This way we keep the query building with the `Processor`. We might even be able to make a generic  `BaseProcessor` with bind cleaning + replacement methods.

**Comment by electricjones** [comment]
I actually came to the same conclusion last night as I was getting ready for bed. I don't mind a Driver relying on its own LanguageProcessor, so long as the end user doesn't have to know what's happening. I never intended for the `Driver` to pass the binds. It should only send completed commands in a specific language and return results. Single responsibility.

For what its worth, `:values` may not work with Orient because the record ids `@rid` are in the format `#12:2` and I feel it would be too easy to regex replace the wrong bit in weird cases.

I am in favor of `{{values}}`. We can give it some options.

`{{keys}}` -> `name, ship`
`{{values}}` -> `Malcom Reynolds, Serenity`
`{{data}}` -> used in `foreach` by the processor in whatever way is appropriate

Maybe even
`{{values(json)}}` -> `{name: "Malcom Reynolds", ship: "Serenity"}}`

Though this may be extra sugar that'll just slow performance

**Comment by dmill-bz** [comment]
I think `{{value}}` works fine as the generic bind format. And I like the idea of the options.

We need to keep in mind though that gremlin for example would not accept this, and gremlin-server will be expecting to do it's own handling of binds. There are several reasons for this:
- gremlin scripts are compiled at runtime then cached. If you run `g.V().has('name', 'michael')` then `g.V().has('name', 'dylan')` you're compiling twice because the scripts are different. If on the other hand you run `g.V().has('name', BIND_NAME)` a dozen times with different binds it will only compile once, then the cached script will be used. And compiling is really taxing.. think anywhere from 30ms to 400ms depending on the query.
- I've found from experience that gremlin does not deal well with backslashes in strings (any level of escaping). Binding these strings makes the pain go away.

In essence sometimes it will probably be best to attach the binds to the `Driver::$client` protocol and let the server handle it. Actually this should probably be the default behavior with any driver that allows for it.

That's what lead me to thinking that each processor should either handle the binding on it's own (for language specifics) or provide a parent class with the expected format and/or formatting callbacks.

**Comment by dmill-bz** [comment]
Adding this as a side note and reminder. But the `GremlinProcessor` should auto-bind any `value` in key/value scenarios.

**Comment by electricjones** [comment]
Great. Sounds like Gremlin is going to be the problem child at every step of the game, lol.

So:
- We bind data to the `Command($script, $language, $binds)`
- The driver either passes this to the `$client` or to the `Processor` to bind the actual data
- Then, the driver executes the query
- We will use {{var}} and a few options.

Does that work? Are we missing something?

**Comment by dmill-bz** [comment]
Sounds like the gist of it.
I think the builder will use {{var}} and the the processor will convert that to whatever is appropriate for the language (if need be)

**Comment by electricjones** [comment]
Considering the builder only produces a `CommandBag`, I would thing the `Processor` would actually use `{{var}}` in its production of `Command`. 

Actually, now that I think of it, why would `Processor` write a script with `{{var}}` and pass it to the `Driver` just to have it passed back and reprocessed with the binds?

I doubt many people will ever even see the actual binding, but its there for those who wish to write manual queries.

**Comment by dmill-bz** [comment]
Well the way I saw it, `Bag`might contain some binds (hard coded by the user in where clauses for example). These would be `{{var}}`. The `LanguageProcessor` would then generate a script with a bind format that would be accepted by a server (`B_VAR1` for example in the case of gremlin, `var1` for neo4j etc.), add this to the `Command` with a `language` a `binds`

And yes it feels like an unnecessary round trip but it allows users to provide their own manual `Command` to the driver. Like say 

``` php
// valid gremlin :
new Command("g.V.has('name', B_VAR)", 'gremlin-groovy', ['B_VAR' => 'michael']);
// instead of invalid gremlin:
new Command("g.V.has('name', {{var}})", 'gremlin-groovy', ['{{var}}' => 'michael']);

```

And it also lets the driver have the final say of how to handle binds. They can either have the processor attach them directly, or send them to the server.

**Comment by electricjones** [comment]
Sounds great. What priority do you think this should have. Right now we have several things going on:
- Refactoring tests #27, #26 
- CypherProcessor #21 
- Settling on Traversal APIs and Trees #19 
- Functional tests
- Driver value binding
- and others that can wait (sessions, ssl, etc).

I'd really like to get something pre-released (v0.3) soon to start using Spider and get some feedback from teams.

Should this stay at 0.4 with traversals, or should it be bumped to 0.3 in your opinion?

**Comment by dmill-bz** [comment]
At this stage it doesn't really matter I think. It's a security feature so as such it's quite important but only in a production environment and we're far from it. It's also not supposed to be a BC breaking change so it can wait, same with SSL.

I totally agree with your order. Is the next released planned for after tests and cypher? That sounds like the correct way of going.

**Comment by electricjones** [comment]
Yes, we need tests and cypher. Then release and focus on top-level `Spider` and traversals, I think. It'd be great to get another person interested who could help out with security stuff. That's one reason I want to get a release out there with proper documentation.

**Comment by electricjones** [comment]
No real target date for the release, just once we have tests and cypher.
